File: Board.py
class Board:
    def __init__(self):
        self.board = [[0] * 7 for _ in range(6)]
        self.turn = 1  # 1 = player 1, 2 = player 2
        self.move = None  # Jogada feita pelo jogador {turn}
        self.parent = None
        self.depth = 0

    def print_board(self):
        print("-----------------------------")
        print("| 1 | 2 | 3 | 4 | 5 | 6 | 7 |")
        print("-----------------------------")
        for row in self.board:
            for cell in row:
                if cell == 1:
                    print("| X ", end="")
                elif cell == 2:
                    print("| O ", end="")
                else:
                    print("|   ", end="")
            print("|")
            print("-----------------------------")

    def valid_move(self, col):
        if col < 0 or col > 6 or self.board[0][col] != 0: #verifica se nao esta a escrever fora do limite ou se ja nao tem uma peça naqula posição
            return False
        return True

    def make_move(self, column):
        if not self.valid_move(column):
            return None
        else:
            new_board = [row[:] for row in self.board]
            for i in range(5, -1, -1):
                if new_board[i][column] == 0:
                    new_board[i][column] = self.turn
                    break
            temp = Board()
            temp.board = new_board
            temp.turn = self.change_turn()
            return temp

    def change_turn(self):
        if self.turn == 1:
            return 2
        return 1

    def is_winner(self):
        if self.evaluator() == 512 or self.evaluator() == -512:
            return True
        return False

    def evaluator(self):
        util = 0
        for h in range(6):
            for u in range(4):
                x = 0
                o = 0
                for i in range(u, u + 4):
                    if self.board[h][i] == 1:
                        x += 1
                    if self.board[h][i] == 2:
                        o += 1
                if x == 4 and o == 0:
                    return 512
                if x == 3 and o == 0:
                    util += 50
                if x == 2 and o == 0:
                    util += 10
                if x == 1 and o == 0:
                    util += 1
                if x == 0 and o == 1:
                    util -= 1
                if x == 0 and o == 2:
                    util -= 10
                if x == 0 and o == 3:
                    util -= 50
                if x == 0 and o == 4:
                    return -512
        for h in range(7):
            for u in range(3):
                x = 0
                o = 0
                for i in range(u, u + 4):
                    if self.board[i][h] == 1:
                        x += 1
                    if self.board[i][h] == 2:
                        o += 1
                if x == 4 and o == 0:
                    return 512
                if x == 3 and o == 0:
                    util += 50
                if x == 2 and o == 0:
                    util += 10
                if x == 1 and o == 0:
                    util += 1
                if x == 0 and o == 1:
                    util -= 1
                if x == 0 and o == 2:
                    util -= 10
                if x == 0 and o == 3:
                    util -= 50
                if x == 0 and o == 4:
                    return -512
        for h in range(3):
            for u in range(4):
                x = 0
                o = 0
                for i in range(4):
                    if self.board[h + i][u + i] == 1:
                        x += 1
                    if self.board[h + i][u + i] == 2:
                        o += 1
                if x == 4 and o == 0:
                    return 512
                if x == 3 and o == 0:
                    util += 50
                if x == 2 and o == 0:
                    util += 10
                if x == 1 and o == 0:
                    util += 1
                if x == 0 and o == 1:
                    util -= 1
                if x == 0 and o == 2:
                    util -= 10
                if x == 0 and o == 3:
                    util -= 50
                if x == 0 and o == 4:
                    return -512
        for h in range(3):
            for u in range(3, 7):
                x = 0
                o = 0
                for i in range(4):
                    if self.board[h + i][u - i] == 1:
                        x += 1
                    if self.board[h + i][u - i] == 2:
                        o += 1
                if x == 4 and o == 0:
                    return 512
                if x == 3 and o == 0:
                    util += 50
                if x == 2 and o == 0:
                    util += 10
                if x == 1 and o == 0:
                    util += 1
                if x == 0 and o == 1:
                    util -= 1
                if x == 0 and o == 2:
                    util -= 10
                if x == 0 and o == 3:
                    util -= 50
                if x == 0 and o == 4:
                    return -512
        return util


class ConnectFour:
    def main(self, mode):
        inicial = Board()

        if mode == "PVP":
            self.play_pvp(inicial)

        elif mode == "MCTS":
            self.play_mcts(inicial)

        else:
            print("Invalid Argument")
            print("Usage: python ConnectFour.py [PVP | MCTS]")

    def play_pvp(self, inicial):
        inicial.print_board()

        for i in range(42):
            print("\n      Player", (i % 2 + 1), "move: ", end="")
            col = int(input()) - 1

            while not inicial.valid_move(col):
                print(" Invalid move, try again: ", end="")
                col = int(input()) - 1

            inicial = inicial.make_move(col)
            inicial.print_board()

            if inicial.is_winner():
                print("Player", (i % 2 + 1), "wins !")
                break


if __name__ == "__main__":
    import sys

    connect_four = ConnectFour()
    connect_four.main(sys.argv[1])

File: Board.java
import java.util.ArrayList;
import java.util.List;

public class Board {
    public int[][] board = new int[6][7];

    int turn; // 1 = player 1, 2 = player 2
    int move; // Jogada feita pelo jogador {turn}
    Board parent;
    private int cost;

    int depth;

    public Board() {
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 7; j++) {
                board[i][j] = 0;
            }
        }

        turn = 1;
    }

    public Board(int[][] mat, int op) {
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 7; j++) {
                board[i][j] = mat[i][j];
            }
        }
        turn = op;
    }

    // --------------------------------------------------------------------------------------

    public void printBoard() {
        System.out.println("-----------------------------");
        System.out.println("| 1 | 2 | 3 | 4 | 5 | 6 | 7 |");
        System.out.println("-----------------------------");
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 7; j++) {
                // imprime as jogadas do jogador 1 como "x" e a joagadas do jogador 2 como "O" e imprime '-' se nao estiver nenhuma peça lá
                if (board[i][j] == 1) {
                    System.out.print("| X ");
                } else if (board[i][j] == 2) {
                    System.out.print("| O ");
                } else {
                    System.out.print("|   ");
                }
            }
            System.out.println("|");
            System.out.println("-----------------------------");
        }
    }

    public int[][] getBoard() {
        return board;
    }
    public void setCost(int cost){
        this.cost=cost;
    }
    public int getCost(){
        return cost;
    }

    public void setDepth(int p) {
        depth = p;
    }

    public int getDepth() {
        return depth;
    }

    public void setMove(int p) {
        move = p;
    }

    public int getMove() {
        return move;
    }

    public void setParent(Board p){ parent = p;}

    public Board getParent(){ return parent;}

    public boolean isTerminal() {
        for (int i = 0; i < 6; i++) {
            if (board[0][i] == 0)
                return false;

        }
        return true;
    }

    public int isFullyExpanded() {
        // Se tem vencedor ou se é empate
        if(evaluator() == 512) return 1;
        if(evaluator() == -512) return 2;
        if(isOver()) return 3;
        return 0;
    }

    public boolean canInsert(int column){
        for (int i = 5; i >= 0; i--) {
            if (board[i][column] == 0) {
                return true;
            }
        }
        return false;

    }

    public int changeTurn() {
        if (turn == 1)
            return 2;
        return 1;
    }

    public int getTurn() {
        return turn;
    }

    public boolean validMove(int col){
        if(col < 0 || col > 6 || board[0][col] != 0) return false;
        return true;
    }

    public Board makeMove(int column) {

        if (!validMove(column)) {
            return null;
        } else {

            if (turn == 1) {
                for (int i = 5; i >= 0; i--) {
                    if (board[i][column] == 0) {
                        board[i][column] = 1;
                        break;
                    }
                }
            } else {
                for (int i = 5; i >= 0; i--) {
                    if (board[i][column] == 0) {
                        board[i][column] = 2;
                        break;
                    }
                }
            }

            Board temp = new Board(board, changeTurn());
            return temp;
        }
    }

    public boolean isOver()
    {
        for (int i = 0; i < 6; i++)
        {
            if(board[0][i] == 0) return false;
        }
        return true;
    }

    public boolean isWinner() {
        if(evaluator() == 512 || evaluator() == -512) return true;
        return false;
    }

    public int evaluator() {

        // Para todos os 4 lados nos vamos fazer a soma da união entre peças iguais para as diagonais, na horizontal e na vertical
        int util = 0;

        // A win by X returns +512 points, a win by O returns -512 points
        // -50 for three Os, no Xs,
        // -10 for two Os, no Xs,
        // - 1 for one O, no Xs,
        // 0 for no tokens, or mixed Xs and Os,
        // 1 for one X, no Os,
        // 10 for two Xs, no Os,
        // 50 for three Xs, no Os

        // Beggining by the horizontals, we need to count 4 lines per board line, hence:
        for (int h = 0; h < 6; h++) {
            for (int u = 0; u < 4; u++) {
                // Count x's and o's on the 4 segment
                int x = 0;
                int o = 0;
                for (int i = u; i < u + 4; i++) {
                    if (board[h][i] == 1)
                        x++;
                    if (board[h][i] == 2)
                        o++;

                }

                // Verifica o resultado e adiciona á soma
                if (x == 4 && o == 0)
                    return 512;
                if (x == 3 && o == 0)
                    util += 50;
                if (x == 2 && o == 0)
                    util += 10;
                if (x == 1 && o == 0)
                    util += 1;
                if (x == 0 && o == 1)
                    util -= 1;
                if (x == 0 && o == 2)
                    util -= 10;
                if (x == 0 && o == 3)
                    util -= 50;
                if (x == 0 && o == 4)
                    return -512;

            }
        }

        // Verticals

        for (int h = 0; h < 7; h++) {
            for (int u = 0; u < 3; u++) {
                // Count x's and o's on the 4 segment
                int x = 0;
                int o = 0;
                for (int i = u; i < u + 4; i++) {
                    if (board[i][h] == 1)
                        x++;
                    if (board[i][h] == 2)
                        o++;

                }

                // Verifica o resultado e adiciona á soma
                if (x == 4 && o == 0)
                    return 512;
                if (x == 3 && o == 0)
                    util += 50;
                if (x == 2 && o == 0)
                    util += 10;
                if (x == 1 && o == 0)
                    util += 1;
                if (x == 0 && o == 1)
                    util -= 1;
                if (x == 0 && o == 2)
                    util -= 10;
                if (x == 0 && o == 3)
                    util -= 50;
                if (x == 0 && o == 4)
                    return -512;

            }
        }

        // Diagonals Pointing Up

        for (int h = 0; h < 3; h++) {
            for (int u = 0; u < 4; u++) {
                // Count x's and o's on the 4 segment
                int x = 0;
                int o = 0;
                for (int i = 0; i < 4; i++) {
                    if (board[h + i][u + i] == 1)
                        x++;
                    if (board[h + i][u + i] == 2)
                        o++;

                }

                // Verifica o resultado e adiciona á soma
                if (x == 4 && o == 0)
                    return 512;
                if (x == 3 && o == 0)
                    util += 50;
                if (x == 2 && o == 0)
                    util += 10;
                if (x == 1 && o == 0)
                    util += 1;
                if (x == 0 && o == 1)
                    util -= 1;
                if (x == 0 && o == 2)
                    util -= 10;
                if (x == 0 && o == 3)
                    util -= 50;
                if (x == 0 && o == 4)
                    return -512;

            }
        }

        // Diagonais para cima

        for (int h = 0; h < 3; h++) {
            for (int u = 3; u < 7; u++) {
                int x = 0;
                int o = 0;
                for (int i = 0; i < 4; i++) {
                    if (board[h + i][u - i] == 1)
                        x++;
                    if (board[h + i][u - i] == 2)
                        o++;

                }

                // Verifica o resultado e adiciona á soma
                if (x == 4 && o == 0)
                    return 512;
                if (x == 3 && o == 0)
                    util += 50;
                if (x == 2 && o == 0)
                    util += 10;
                if (x == 1 && o == 0)
                    util += 1;
                if (x == 0 && o == 1)
                    util -= 1;
                if (x == 0 && o == 2)
                    util -= 10;
                if (x == 0 && o == 3)
                    util -= 50;
                if (x == 0 && o == 4)
                    return -512;

            }
        }

        return util;
    }

}
